src/connection.rs

```diff
@@ -46,29 +46,44 @@ impl Connection {
         Ok(self.transformer.deserialize()?.0)
     }
 
-    async fn read_and_deserialize<'a, T>(&'a mut self, size: usize) -> Result<T>
-    where
-        T: Deserialize<'a>,
-    {
-        self.transformer.get_buffer().clear();
-        read_exact_to_vec(
-            &mut self.raw_conn.stream,
-            self.transformer.get_buffer(),
-            size,
-        )
-        .await?;
+    pub(crate) async fn read_response(&mut self) -> Result<Response> {
+        // Clear content of buffer
+        //
+        // We assume that previous call to `Connection::read_response` does not
+        // read any bytes of the current packet to read in.
+        let buffer = self.transformer.get_buffer();
+        buffer.clear();
 
-        self.deserialize()
-    }
+        // Deserialize len of the packet (first 4b)
+        //
+        // We will read in at most 16 bytes here to cache the whole packet
+        // because any response takes at least 16 bytes
+        //  - 0..4 len of packet
+        //  - 4..8 type of the packet
+        //  - 8..16 version (`Response::Hello`), `response_id` or `session_id`.
+        //
+        // Reading more than that will break precondition of next call to
+        // `Connection::read_response`.
+        read_to_vec_rng(&mut self.raw_conn.stream, buffer, 4..16).await?;
 
-    /// Return size of the response.
-    async fn read_header(&mut self) -> Result<u32> {
-        self.read_and_deserialize(4).await
-    }
+        let len: u32 = self.deserialize()?;
 
-    pub(crate) async fn read_response(&mut self) -> Result<Response> {
-        let len = self.read_header().await?;
-        self.read_and_deserialize(len as usize).await
+        // Remove first 4b representing the len
+        let buffer = self.transformer.get_buffer();
+        buffer.drain(..4);
```

Comment:
What's the type of `buffer` here? If it's not a `Bytes`, this could be quite expensive.

Reply:
It is a Vec<u8> and that is indeed a bit expensive.

Reply:
I guess it might be better to use `BytesMut` here to share the buffer between `Serializer` and `Deserializer` while making `drain` O(1).

Reply:
I've fixed this by doing one `drain` before returning from `read_response`.